libraries/apollo-runtime/src/appleMain/kotlin/com/apollographql/apollo3/network/http/DefaultHttpEngine.apple.kt

```diff
@@ -320,3 +303,20 @@ private class Pipe {
     }
   }
 }
+
+fun interface DataTaskFactory {
+  fun dataTask(request: NSURLRequest, delegate: NSURLSessionDataDelegateProtocol): NSURLSessionDataTask
+}
+
+private class DefaultDataTaskFactory : DataTaskFactory {
+  private var nsUrlSession: NSURLSession? = null
+
+  override fun dataTask(request: NSURLRequest, delegate: NSURLSessionDataDelegateProtocol): NSURLSessionDataTask {
+    if (nsUrlSession == null) nsUrlSession = NSURLSession.sessionWithConfiguration(
+        configuration = NSURLSessionConfiguration.defaultSessionConfiguration(),
+        delegate = delegate,
+        delegateQueue = null
+    )
+    return nsUrlSession!!.dataTaskWithRequest(request)
```

Comment:
This is prone to race conditions and also surprising because `delegate` would only be taken into account for the first call. What about

```kotlin
fun interface DataTaskFactory {
  fun dataTask(request: NSURLRequest): NSURLSessionDataTask
}

private class DefaultDataTaskFactory(private val delegate) : DataTaskFactory {
  // ...
}
```

Reply:
Also more general comment about `fun interface`. We don't have many in the codebase and I think I'd prefer not adding too many. Regular interfaces are probably easier to evolve if we need to

Reply:
Ah but wait, we need to control the delegate so looks like we need a factory of factory instead 🙈